platform/core/src/services/DisplaySetService/DisplaySetService.ts

```diff
@@ -46,42 +62,63 @@ export default class DisplaySetService extends PubSubService {
     this.extensionManager = extensionManager;
     this.SOPClassHandlerIds = SOPClassHandlerIds;
     this.activeDisplaySets = [];
+    this.activeDisplaySetsMap.clear();
   }
 
-  _addDisplaySetsToCache(displaySets) {
+  _addDisplaySetsToCache(displaySets: DisplaySet[]) {
     displaySets.forEach(displaySet => {
-      displaySetCache.push(displaySet);
+      displaySetCache.set(displaySet.displaySetInstanceUID, displaySet);
     });
   }
 
-  _addActiveDisplaySets(displaySets) {
-    const activeDisplaySets = this.activeDisplaySets;
+  _addActiveDisplaySets(displaySets: DisplaySet[]) {
+    const { activeDisplaySets, activeDisplaySetsMap } = this;
 
     displaySets.forEach(displaySet => {
-      // This test makes adding display sets an N^2 operation, so it might
-      // become important to do this in an efficient manner for large
-      // numbers of display sets.
-      if (!activeDisplaySets.includes(displaySet)) {
+      if (!activeDisplaySetsMap.has(displaySet.displaySetInstanceUID)) {
+        this.activeDisplaySetsChanged = true;
         activeDisplaySets.push(displaySet);
+        activeDisplaySetsMap.set(displaySet.displayInstanceUID, displaySet);
       }
     });
   }
 
-  getDisplaySetCache() {
+  /**
+   * Adds new display sets directly, as specified.
+   * Use this function when the display sets are created externally directly
+   * rather than using the default sop class handlers to create display sets.
+   */
+  public addDisplaySets(...displaySets: DisplaySet[]): string[] {
+    this._addDisplaySetsToCache(displaySets);
+    this._addActiveDisplaySets(displaySets);
+
+    // The activeDisplaySetsChanged flag is only seen if we add display sets
```

Comment:
I am not sure if the comment still applies or perhaps I am missing something or maybe I just do not understand the comment. Please clarify.